sys/net/gnrc/network_layer/ipv6/ext/gnrc_ipv6_ext.c

```diff
@@ -25,53 +30,174 @@
-        return GNRC_IPV6_EXT_RH_AT_DST;
+gnrc_pktsnip_t *gnrc_ipv6_ext_process_all(gnrc_pktsnip_t *pkt,
+                                          uint8_t *nh)
+{
+    bool is_ext = true;
+    while (is_ext) {
+        switch (*nh) {
+            case PROTNUM_IPV6_EXT_DST:
+            case PROTNUM_IPV6_EXT_RH:
+            case PROTNUM_IPV6_EXT_FRAG:
+            case PROTNUM_IPV6_EXT_AH:
+            case PROTNUM_IPV6_EXT_ESP:
+            case PROTNUM_IPV6_EXT_MOB: {
+                ipv6_ext_t *ext_hdr;
+
+                DEBUG("ipv6: handle extension header (nh = %u)\n", *nh);
+                ext_hdr = pkt->data;
+                if ((pkt = _demux(pkt, *nh)) == NULL) {
+                    DEBUG("ipv6: packet was consumed by extension header "
+                          "handling\n");
+                    return NULL;
+                }
+                *nh = ext_hdr->nh;
+                if (_duplicate_hopopt(pkt, *nh)) {
+                    return NULL;
+                }
+                break;
+            }
+            default:
+                is_ext = false;
+                break;
+        }
     }
+    return pkt;
+}
 
-    /* We cannot use `gnrc_pktbuf_start_write` since it duplicates only
-       the head. `ipv6_ext_rh_process` modifies the IPv6 header as well as
-       the extension header */
+static bool _duplicate_hopopt(gnrc_pktsnip_t *pkt, unsigned nh)
+{
+    if (nh == PROTNUM_IPV6_EXT_HOPOPT) {
+        DEBUG("ipv6: duplicate Hop-by-Hop header\n");
+#ifdef MODULE_GNRC_ICMPV6_ERROR
+        gnrc_icmpv6_error_param_prob_send(
+                ICMPV6_ERROR_PARAM_PROB_NH,
+                &((ipv6_ext_t *)pkt->next->data)->nh, pkt
+            );
+#endif
+        gnrc_pktbuf_release_error(pkt, EINVAL);
+        return true;
+    }
+    return false;
+}
 
-    current_offset = gnrc_pkt_len_upto(current->next, GNRC_NETTYPE_IPV6);
+#ifdef MODULE_GNRC_IPV6_EXT_RH
+/* unchecked precondition: hdr is gnrc_pktsnip_t::data of the
+ * GNRC_NETTYPE_IPV6 snip within pkt */
+static void _forward_pkt(gnrc_pktsnip_t *pkt, ipv6_hdr_t *hdr)
+{
+    gnrc_pktsnip_t *netif_snip;
 
-    if (pkt->users != 1) {
-        if ((ipv6 = gnrc_pktbuf_duplicate_upto(pkt, GNRC_NETTYPE_IPV6)) == NULL) {
-            DEBUG("ipv6: could not get a copy of pkt\n");
-            gnrc_pktbuf_release(pkt);
-            return GNRC_IPV6_EXT_RH_ERROR;
-        }
-        pkt = ipv6;
-        hdr = ipv6->data;
-        ext = (ipv6_ext_t *)(((uint8_t *)ipv6->data) + current_offset);
+    if (--(hdr->hl) == 0) {
+        DEBUG("ipv6_ext_rh: hop limit reached 0: drop packet\n");
+        gnrc_pktbuf_release(pkt);
     }
-    else {
-        ipv6 = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_IPV6);
-        hdr = ipv6->data;
+    /* remove L2 headers around IPV6 */
+    netif_snip = gnrc_pktsnip_search_type(pkt, GNRC_NETTYPE_NETIF);
+    if (netif_snip != NULL) {
+        pkt = gnrc_pktbuf_remove_snip(pkt, netif_snip);
+    }
+    pkt = gnrc_pktbuf_reverse_snips(pkt);
```

Comment:
mayyyybe add a comment on why they need to be reversed?

Reply:
I will add a comment in #10238.

Reply:
See https://github.com/RIOT-OS/RIOT/pull/10229/commits/971f1d0fa07010187f79c2616ccf1253c4bced3c (and https://github.com/RIOT-OS/RIOT/pull/10229#issuecomment-441188292).